toplevel/opttoploop.ml

```diff
@@ -455,20 +461,25 @@ let use_channel ppf ~wrap_in_module ic name filename =
   Lexer.skip_hash_bang lb;
   let success =
     protect_refs [ R (Location.input_name, filename) ] (fun () ->
+      let log = Location.init_log ppf in
       try
         List.iter
           (fun ph ->
-            let ph = preprocess_phrase ppf ph in
-            if not (execute_phrase !use_print_results ppf ph) then raise Exit)
+            let ph = preprocess_phrase log  ph in
+            if not (execute_phrase !use_print_results log ph) then raise Exit)
           (if wrap_in_module then
              parse_mod_use_file name lb
            else
              !parse_use_file lb);
+        Misc.Log.flush log;
         true
       with
-      | Exit -> false
-      | Sys.Break -> fprintf ppf "Interrupted.@."; false
-      | x -> Location.report_exception ppf x; false) in
+      | Exit -> Misc.Log.flush log; false
+      | Sys.Break ->
+          Misc.Log.logf "status" log "Interrupted.@.";
+          Misc.Log.flush log;
+          false
+      | x -> Location.report_exception log x;   Misc.Log.flush log; false) in
```

Comment:
we would want to factorize the three `flush` calls here.

Reply:
(we could use a function `Location.with_log : formatter -> (unit -> 'a) -> 'a`)

Reply:
I have factorized the code with a `with_log` function as much as possible (there is at least one case when we want to avoid printing empty object needlessly in the toplevel).